### ISSUE TITLE
overload std::ostream& operator<<(std::ostream&, typename T) causes gcc compile failure on release v.1.11.0

### ISSUE THREAD

**Issue by ice10513** [google/googletest#3552]
**Describe the bug**

I saw a similar bug already reported at [issue-3384](https://github.com/google/googletest/issues/3384), but the problem not completely the same.

```
Output of ARM gcc 11.1 (linux) (Compiler #1)
In file included from <source>:6:
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h: In instantiation of 'struct testing::internal::FindFirstPrinter<Detail::MyValue, void, testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter, testing::internal::ProtobufPrinter, testing::internal::ConvertibleToIntegerPrinter, testing::internal::ConvertibleToStringViewPrinter, testing::internal::RawBytesPrinter, testing::internal::FallbackPrinter>':
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:285:8:   recursively required from 'struct testing::internal::FindFirstPrinter<Detail::MyValue, void, testing::internal::FunctionPointerPrinter, testing::internal::PointerPrinter, testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter, testing::internal::ProtobufPrinter, testing::internal::ConvertibleToIntegerPrinter, testing::internal::ConvertibleToStringViewPrinter, testing::internal::RawBytesPrinter, testing::internal::FallbackPrinter>'
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:285:8:   required from 'struct testing::internal::FindFirstPrinter<Detail::MyValue, void, testing::internal::ContainerPrinter, testing::internal::FunctionPointerPrinter, testing::internal::PointerPrinter, testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter, testing::internal::ProtobufPrinter, testing::internal::ConvertibleToIntegerPrinter, testing::internal::ConvertibleToStringViewPrinter, testing::internal::RawBytesPrinter, testing::internal::FallbackPrinter>'
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:305:9:   required from 'void testing::internal::PrintWithFallback(const T&, std::ostream*) [with T = Detail::MyValue; std::ostream = std::basic_ostream<char>]'
<source>:29:41:   required from here
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:289:36: error: no matching function for call to 'testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const Detail::MyValue&, std::nullptr_t)'
  289 |     T, decltype(Printer::PrintValue(std::declval<const T&>(), nullptr)),
      |                 ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:210:15: note: candidate: 'template<class T, class, class> static void testing::internal::internal_stream_operator_without_lexical_name_lookup::StreamPrinter::PrintValue(const T&, std::ostream*)'
  210 |   static void PrintValue(const T& value, ::std::ostream* os) {
      |               ^~~~~~~~~~
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:210:15: note:   template argument deduction/substitution failed:
/opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:209:33: error: no match for 'operator<<' (operand types are 'std::basic_ostream<char>' and 'const Detail::MyValue')
  208 |             typename = decltype(std::declval<std::ostream&>()
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  209 |                                 << std::declval<const T&>())>
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/compiler-explorer/libs/googletest/trunk/googletest/include/gtest/gtest-printers.h:103,
                 from <source>:6:
/opt/compiler-explorer/arm/gcc-11.1.0/arm-unknown-linux-gnueabihf/arm-unknown-linux-gnueabihf/include/c++/11.1.0/ostream:108:7: note: candidate: 'std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::basic_ostream<_CharT, _Traits>::__ostream_type& (*)(std::basic_ostream<_CharT, _Traits>::__ostream_type&)) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]'
  108 |       operator<<(__ostream_type& (*__pf)(__ostream_type&))
      |       ^~~~~~~~
```

**Steps to reproduce the bug**

[https://godbolt.org/z/YaP7zafPM](https://godbolt.org/z/YaP7zafPM)

Code:
```cpp
#include <gtest/gtest-printers.h>

struct MyData
{
    template <typename T>
    MyData(T) {}
};

std::ostream& operator<<(std::ostream& os, const MyData&)
{
    return os;
}

namespace Detail
{
struct MyValue{};
}

void foo()
{
    testing::internal::PrintWithFallback(Detail::MyValue{}, &std::cout);
}
```

**Does the bug persist in the most recent commit?**

I encountered the problem when I upgrade from release v.1.10.0 to v.1.11.0.

**What operating system and version are you using?**

N/A

**What compiler and version are you using?**

gcc10.2, but I tested it on the Godbolt, the clang can success compile it, but GCC cannot, not sure who's the problem.

**What build system are you using?**
N/A

**Additional context**

**Comment by asoffer** [comment]
Thanks for letting us know about this. It looks like this is fixed in GCC at trunk and in Clang reasonably far back, so we're not going to provide a workaround inside the GoogleTest library.

That being said, it looks making the constructor explicit and/or making `operator<<` a hidden friend reduce the lookup scope sufficiently to work around the issue. https://godbolt.org/z/aYPznYccY

**Comment by Finkman** [comment]
The recommenced work around is not possible for enums. I am struggling to build with gcc 11 against latest gtest.
I am not sure if it is a wise decision to not fix this inside the gtest lib, since it is failing with default compilers (gcc 11) which are shipped on some LTS distris (Ubuntu 22.04).

**Comment by dev0x13** [comment]
Pending PR fixing this (for visibility): https://github.com/google/googletest/pull/3993
